compiler/src/dotty/tools/dotc/typer/Linter.scala

```diff
@@ -0,0 +1,126 @@
+      case _ => false
+
+    def checkInterestingShapes(t: Tree): Boolean = t match
+      case If(_, thenpart, elsepart) => checkInterestingShapes(thenpart) || checkInterestingShapes(elsepart)
+      case Block(_, res) => checkInterestingShapes(res)
+      case Match(_, cases) => cases.exists(k => checkInterestingShapes(k.body))
+      case _ => checksForInterestingResult(t)
+
+    def checksForInterestingResult(t: Tree): Boolean =
+         !t.isDef                               // ignore defs
+      && !isUninterestingSymbol(t.symbol)       // ctors, package, Unit, Any
+      && !isUninterestingType(t.tpe)            // bottom types, Unit, Any
+      && !isThisTypeResult(t)                   // buf += x
+      && !isSuperConstrCall(t)                  // just a thing
+      && !isJavaApplication(t)                  // Java methods are inherently side-effecting
+      // && !treeInfo.hasExplicitUnit(t)           // suppressed by explicit expr: Unit // TODO Should explicit `: Unit` be added as warning suppression?
+
+    if ctx.settings.WNonUnitStatement.value && !ctx.isAfterTyper && checkInterestingShapes(t) then
+      val where = t match
+        case Block(_, res) => res
+        case If(_, thenpart, Literal(Constant(()))) =>
+          thenpart match {
+            case Block(_, res) => res
+            case _ => thenpart
+          }
+        case _ => t
+      report.warning(UnusedNonUnitValue(where.tpe), t.srcPos)
+      true
+    else false
+  end warnOnInterestingResultInStatement
+
+  /** If -Wimplausible-patterns is set, warn about pattern values that can match the scrutinee
+   *  type only if there would be some user-defined equality method that equates values of the
+   *  two types.
+   */
+  def warnOnImplausiblePattern(pat: Tree, selType: Type)(using Context): Unit =
+      // approximate type params with bounds
+    def approx = new ApproximatingTypeMap {
+      var alreadyExpanding: List[TypeRef] = Nil
+      def apply(tp: Type) = tp.dealias match
+        case tp: TypeRef if !tp.symbol.isClass =>
+          if alreadyExpanding contains tp then tp else
+            val saved = alreadyExpanding
+            alreadyExpanding ::= tp
+            val res = expandBounds(tp.info.bounds)
+            alreadyExpanding = saved
+            res
+        case _ =>
+          mapOver(tp)
+    }
+
+    // Is it possible that a value of `clsA` is equal to a value of `clsB`?
+    //  This ignores user-defined equals methods, but makes an exception
+    //  for numeric classes.
+    def canOverlap(clsA: ClassSymbol, clsB: ClassSymbol): Boolean =
+      clsA.mayHaveCommonChild(clsB)
+      || clsA.isNumericValueClass // this is quite coarse, but matches to what was done before
+      || clsB.isNumericValueClass
+
+    // Can type `a` possiblly have a common instance with type `b`?
+    def canEqual(a: Type, b: Type): Boolean = trace(i"canEqual $a $b"):
```

Comment:
Minor: There is `TypeComparer.provablyDisjoint`, it seems to be good enough for the purpose of linting --- the semantics is a little different though.

Reply:
I looked at that, but it seems different enough to be easily adaptable. And I don't want linter code to complicate the other code. So if we could use provably disjoint as is, it would be OK. But it seems the only point where we could use it is to implement the `canHaveCommonChild` functionality, and that one is simpler by itself.